docs/DafnyRef/Expressions.md

```diff
@@ -584,40 +584,87 @@ or `this`, which denotes the current object in the context of
 an instance method or function.
 
 ## 20.21. Fresh Expressions {#sec-fresh-expression}
-````grammar
-FreshExpression_ =
-  "fresh" [ "@" LabelName ]
-  "(" Expression(allowLemma: true, allowLambda: true) ")"
-````
 
 `fresh(e)` returns a boolean value that is true if
 the objects denoted by expression `e` were all
-freshly allocated since the time of entry to the enclosing method.
+freshly allocated since the time of entry to the enclosing method,
+or since [`label L:`](#sec-labeled-stmt) in the variant `fresh@L(e)`.
+For example, the following program verifies:
+
+```dafny
+class C { constructor() {} }
+method f(x1: C) returns (x: C)
+  ensures fresh(x)
+{
+  assert !fresh(x1);
+  var x2 := new C();
+  label AfterX2:
+  var x3 := new C();
+  assert fresh(x2) && fresh(x3);
+  assert fresh({x2, x3});
+  assert !fresh@AfterX2(x2) && fresh@AfterX2(x3);
+  x := x2;
+}
+```
 
-If the `LabelName` is present, it must denote a label that in the
-enclosing method's control flow dominates the expression. In this
+The `L` in the variant `fresh@L(e)` must denote a [label](#sec-labeled-stmt) that, in the
+enclosing method's control flow, [dominates the expression](#sec-labeled-stmt). In this
 case, `fresh@L(e)` returns `true` if the objects denoted by `e` were all
 freshly allocated since control flow reached label `L`.
 
-The argument of `fresh` must be either an object reference
+The argument of `fresh` must be either a non-null object reference
 or a set or sequence of object references.
+In this case, `fresh(e)` (respectively `fresh@L(e)` with a label)
```

Comment:
```suggestion
In any case, `fresh(e)` (respectively `fresh@L(e)` with a label)
```

Reply:
fresh(e) does not accept datatypes like `!old(allocated(e))` would. So it's really in this case. The following verifies:
```dafny
class C {
  constructor(){}
}

datatype List = Nil | Cons(head: C, List)

method f() returns (l: List)
  ensures !old(allocated(l))
{
  var x := new C();
  l := Cons(x, Nil);
}
```

But the following does not ("the argument of a fresh expression must denote an object or a set or sequence of objects (instead got List)")
```dafny
class C {
  constructor(){}
}

datatype List = Nil | Cons(head: C, List)

method f() returns (l: List)
  ensures fresh(l)
{
  var x := new C();
  l := Cons(x, Nil);
}